Reverse intensity (Insight)

dominikl opened the pull request:
# What this PR does

Adds the reverse intensity functionality to Insight.
# Testing this PR

In preview and full viewer: Open the color picker for a channel, reverse the intensity by ticking the checkbox in the upper right corner of the color picker.

dominikl commented:
By the way, I can move the checkbox/button somewhere else, if needed, not much hassle. Couldn't check how it looks like in OMERO.web, as #4839 apparently isn't in the build.

gusferguson commented:
Tested using OMERO.insight-5.3.0-m4-253-43428df-ice36-b434-win eel user-3 read-only-1 image ID 24763

Location of checkbox appears fine.
Clicking the reverse intensity checkbox in the color picker appears to have the desired effect for both LUTs and plain colours and Preview and Accept buttons function as expected.

But:

Open image in Full Viewer - open Rendering Setting panes
- as expected most of time - intermittently getting black user settings thumbnail in Full Viewer Rendering Settings pane - thumbnails in centre pane and Preview pane as expected - can’t reproduce reliably - may be after a number of channel colour changes and saves - see screenshot 1

Open image with imported settings in Full Viewer - open Rendering Settings pane
- thumbnails and Preview as expected
- use color picker on 3rd channel to select LUT - Green fire blue - click accept
- thumbnail in FV RS pane user settings remains unchanged
- thumbnail in Preview pane user settings - black
- thumbnail in centre pane - unchanged
- click save - no changes - screenshot 2
- close FV - click refresh in Data Tree - still no update of thumbnails in centre pane or user settings in Preview pane

Re-open same image in FV - open RS pane
- select color picker on same third channel - select reverse intensity - click accept
- FV image reflects reversed intensity - Save button does not enable
- Preview pane does not change
- click on main window to get focus - Preview pane updates to show reverse intensity - save not enabled
- close full viewer - select another image - go back to previous image - reverse intensity setting has not been saved
- screenshots 3 and 4

Log out and log back in - LUT (not reversed intensity) showing in Preview pane and FV but all thumbnails in centre pane and user settings still show import settings - screenshot 5.

![parallels desktop](https://cloud.githubusercontent.com/assets/1552703/18664463/91f51f1e-7f1a-11e6-9fc4-8dffe7717bee.png)

![parallels desktop](https://cloud.githubusercontent.com/assets/1552703/18664468/9a03cbec-7f1a-11e6-8c88-854b50bc048b.png)

![parallels desktop](https://cloud.githubusercontent.com/assets/1552703/18664478/a20a0a72-7f1a-11e6-9399-f3e53866d6c0.png)

![parallels desktop](https://cloud.githubusercontent.com/assets/1552703/18664484/a70901b8-7f1a-11e6-89ae-e8307ecda947.png)

![parallels desktop](https://cloud.githubusercontent.com/assets/1552703/18664488/acc508f4-7f1a-11e6-9446-e6c9dee8835d.png)

dominikl commented:
Fixed:
- Reverse intensity change wasn't recognized as rendering settings modification, that's why the save button wasn't enabled.
- Various issues with the color picker when used in full image viewer.

Still don't know where the black thumbnail issues come from, couldn't replicate them either.

jburel commented:
@dominikl: few more commits pushed to https://github.com/openmicroscopy/openmicroscopy/pull/4823

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/imviewer/view/ImViewerComponent.java:
```diff
@@ -1082,6 +1082,33 @@ public void setLookupTable(int index, String lut, boolean preview) {
+        case DISCARDED:
+            throw new IllegalStateException(
+                    "This method can't be invoked in the DISCARDED, "
+                            + "NEW or LOADING_RENDERING_CONTROL state.");
```
I assume it's okay that the message mentions `LOADING_RENDERING_CONTROL` but the `switch` doesn't.

dominikl replied:
That's a good point. There are actually quite a lot similar cases where the message of the exception is not really correct. Guess that just slipped in with copy/pastes. Might be worth a separate general clean-up PR. Wouldn't add it to this PR, already rather long and reviewed.

mtbc replied:
It's fine with me if you prefer to fix these in a separate PR, just pop it on a 5.3 card.

jburel replied:
This is more legacy, I will handle it in a separate PR

dominikl replied:
Just for reference: [Trello card - Insight code cleanup](https://trello.com/c/cRMFECqd/128-insight-code-cleanup)

jburel commented:
issue noticed this evening while looking at the black thumbnail problem.
I start with an image with 2 channels: first is green, second red
 * Change first to a lut, save etc.
 * close the viewer
 * re-open the viewer
 * try to reset the color to green. Click accept in the color picker. 
 * The image does not get updated. Color are be compared and are the same but in that case a lut was applied. So the "isSameSettings" need to be reviewed.

jburel commented:
Since the reverse intensity box is in the color picker
When selected, we should probably change the color of the button to match what is happening
Thoughts? cc @will-moore @gusferguson

gusferguson commented:
>Since the reverse intensity box is in the color picker when selected, we should probably change the color of the button to match what is happening (i.e. "reverse color")

Not sure about this - does the button represent the LUT selected and the reverse intensity is something else - or is the reverse intensity the equivalent of a different LUT?

dominikl commented:
Yes, can replicate the problem https://github.com/openmicroscopy/openmicroscopy/pull/4850#issuecomment-248746959 . Thanks for the workflow.

RE https://github.com/openmicroscopy/openmicroscopy/pull/4850#issuecomment-248748544 : Was thinking about that, too. I'd also suggest to reverse/flip the color icon in the channel button (and slider?).

jburel commented:
@gusferguson 
the reverse intensity does ``new_value = 255-value`` e.g. 10 will become 245. So yes it can be seen as a different color/lut

gusferguson commented:
@dominikl 

Tested using OMERO.insight-5.3.0-m4-389-2cbfd1a-ice35-b438-mac eel user-3

Seeing no difference in behaviour at all. 
What exactly was the bug the last commit addressed?

gusferguson commented:
@dominikl 
Tested using OMERO.insight-5.3.0-m4-389-2cbfd1a-ice35-b438-mac eel user-3
Sorry I misunderstood the comment! 
Yes the bug-fix has worked - last commit behaves as expected.

dominikl commented:
Thanks @gusferguson . Yes, the bug was switching from color to lut and back didn't work. I guess the black thumbnail problem is still there, still a bit a mystery to me.

jburel commented:
@dominikl: black thumbnail could be due to a race condition.
Anything in the logs?

jburel commented:
discussed with @dominikl: optimistic lock reported previously see https://trello.com/c/9iAe3Sg9/95-lut-support

Anything more @gusferguson?

jburel commented:
@pwalczysko: could you do a final round of review on this PR and the corresponding Web one i.e. https://github.com/openmicroscopy/openmicroscopy/pull/4839?
server has been merged, and it will be good to wrap up this work

pwalczysko commented:
Bug: 
- Open an image in Full viewer.
-  Go to another image from the same Dataset. Open Preview.
- In Preview, open a channel and check the "Inverse" checkbox, do not save
- Copy the Rnd settings
- go to another (third) image in Preview, paste the Rnd settings, then Save the Rnd settings
- Still on the third image, click Save to All -> okay the dialog and observe that
- all the thumbs in central pane change as expected, excelpt for the image whose full viewer is open

pwalczysko commented:
The default size of thumbs in central pane after you open Insight has changed (became bigger).
![screen shot 2016-10-12 at 14 01 22](https://cloud.githubusercontent.com/assets/2478303/19310979/7ad73184-9084-11e6-90db-17232c668328.png)

This comment is linke to the SPW PR as this is the culprit.

jburel commented:
@dominikl: note that the backend PR has now been merged.

jburel commented:
larger thumbnail has been introduced by https://github.com/openmicroscopy/openmicroscopy/pull/4874/commits/926479fc9ef5e4bf3054a3dd4d18931a685e5446 see PR https://github.com/openmicroscopy/openmicroscopy/pull/4874

pwalczysko commented:
@jburel Okay, ta, made a comment on the SPW PR.

pwalczysko commented:
In summary, one Bug detected https://github.com/openmicroscopy/openmicroscopy/pull/4850#issuecomment-253206140. @dominikl is telling me this is hard to fix, the Bug, although repeatable is a slightly edge workflow case, so not a blocker if cannot be debugged.

Otherwise than that it works nicely.

jburel commented:
Is this edge case also in the release version?

pwalczysko commented:
@jburel Yes, the edge case (=bug) is present in 5.2.5 release insight (connected to demo).

jburel commented:
card for bug https://trello.com/c/X3pLD781/131-bug-insight

dominikl commented:
Rebased to develop, after #4823 has been merged. Probably a quick re-review necessary.

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/util/ui/colourpicker/RGBModel.java:
```diff
@@ -76,50 +77,71 @@
      * @param v value.
      */
-    private void setV(float v)
+    private void setV(float v, boolean reset)
```
javadoc to update

mtbc replied:
There's an Eclipse setting somewhere that can be turned on to make these Javadoc issues yellow warnings. (If you find it, show @pwalczysko!)

dominikl replied:
The default is that the javadoc is only checked for `public` methods; can be changed to `private` here:
![screen shot 2016-10-19 at 11 34 41](https://cloud.githubusercontent.com/assets/6575139/19515662/45ed1cf6-95f0-11e6-92b9-4be55a046524.png)

mtbc replied:
:+1: Maybe our developer docs should say to check "validate tag arguments" and consider members as visible if they are whatever the level _above_ private is (don't think we include private in docgen).

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/util/ui/colourpicker/RGBModel.java:
```diff
@@ -76,50 +77,71 @@
      * @param h hue.
      */
-    private void setH(float h)
+    private void setH(float h, boolean reset)
```
javadoc

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/util/ui/colourpicker/RGBModel.java:
```diff
@@ -76,50 +77,71 @@
      * @param s saturation.
      */
-    private void setS(float s)
+    private void setS(float s, boolean reset)
```
same

pwalczysko commented:
Restest went well, the thumb size as expected, other features either as expected or as reported (e.g. https://github.com/openmicroscopy/openmicroscopy/pull/4850#issuecomment-253740446).

Ready to merge.

will-moore commented:
Noticed a minor bug in today's build...
Launching the color-picker from the icon above the channel buttons to the left of the viewer, the Reverse intensity checkbox is not getting checked/unchecked according to the status of the picked channel.
However, if I launch the color-picker from the main Settings tab then I don't see this bug.

dominikl commented:
Only fixed the javadoc issues. Couldn't replicate https://github.com/openmicroscopy/openmicroscopy/pull/4850#issuecomment-254452606 .

dominikl commented:
Sorry @mtbc . Rebased and force pushed, so now your comment seems to be vanished :-/

jburel commented:
Re-Listing this PR for a final review

pwalczysko commented:
I am afraid I can reproduce https://github.com/openmicroscopy/openmicroscopy/pull/4850#issuecomment-254452606 (the launch of the color picker from LHP of the Full Viewer does not match with the launch of color picker from Settings in Full viewer) as reported by @will-moore .
- login as user-4 to eel
- open a leica image in Full Viewer
- open the Settings tab (in Full Viewer)
- launch a channel color picker from Settings tab and tick the Reverse Intensity box
- confirm, accept and save
- with the Full Viewer still open, now select the color picker from the left-hand pane of the Full Viewer
- select the channel which you have just reversed the intensity on and observe
- the box of reverse intensity is not ticked -> not expected

pwalczysko commented:
Other than https://github.com/openmicroscopy/openmicroscopy/pull/4850#issuecomment-255979849 all works fine.

pwalczysko commented:
Sorry, noticed another, much worse bug today: 
- select a well in a plate, tick on it a reverse intensity checkbox on one channel
- Click Save to All, observe you get the warning dialog
- okay the warning dialog, observe you get another warning dialog again (the same dialog the second time) -> not expected

![screen shot 2016-10-26 at 10 53 17](https://cloud.githubusercontent.com/assets/2478303/19721566/b6194c5a-9b6a-11e6-86c2-b9d9030766ba.png)

dominikl commented:
https://github.com/openmicroscopy/openmicroscopy/pull/4850#issuecomment-256301864 not introduced by this PR (already present in latest build), added to card https://trello.com/c/X3pLD781/131-bug-insight-rnd-settings-copy-paste .

https://github.com/openmicroscopy/openmicroscopy/pull/4850#issuecomment-254452606 : Rebased the PR to current develop branch, can't replicate the issue. Will have to check again tomorrow's merge build.

pwalczysko commented:
The bug as written https://github.com/openmicroscopy/openmicroscopy/pull/4850#issuecomment-255979849 acutally works now. A new hitch found though : if you change the status of the Reverse Intensity on one channel in Settings tab of Full viewer and save and then go to the left-hand side color picker and check any other channel status, it is ticked as well (although the Settings tab shows that it is not.)

dominikl commented:
Thanks @pwalczysko . Hope I fixed it now with the last commit.

pwalczysko commented:
Bug fixed. Ready to merge.

will-moore commented:
Hi - Sorry, not sure if this PR is responsible, but I just noticed that when I right-click and pick new color for channel buttons beside the image, the button goes white. E.g. after picking DAPI -> Green:

![screen shot 2016-11-02 at 10 25 07](https://cloud.githubusercontent.com/assets/900055/19925252/1352525a-a0e7-11e6-8cc7-5729baef08a4.png)

Also, if I pick a new color using the Settings panel this is not updated in the Channel buttons beside the image (no change).

pwalczysko commented:
Yes, confirming @will-moore ' s [observation](https://github.com/openmicroscopy/openmicroscopy/pull/4850#issuecomment-257827449). Definitely a regression. Thanks @will-moore

dominikl commented:
Spotted (and fixed) another bug. If the channel color is changed from a lookup table to a color, it wasn't set this color but it was set to the color which was specified before the lookup table has been set. I don't think both are related to this PR, but makes sense to fix them here. 

I.e. to test last two commits:
- Follow @will-moore 's workflow https://github.com/openmicroscopy/openmicroscopy/pull/4850#issuecomment-257827449 ; make sure it's fixed now.
- Set a channel to a specific color (remember that color); save rendering settings; set channel to a lookup table; save; now set channel to a different color from the previous one; make sure the channel is set to this color and not the one from the initial step.

jburel commented:
Tested the last 2 commits, both problems are now fixed

jburel commented:
cf. RFE https://github.com/openmicroscopy/openmicroscopy/pull/4850#issuecomment-248848806
this is something to probably discuss as a follow. I will add it to the card i.e. https://trello.com/c/SrcTVbw3/135-lut-improvements

One small problem noticed:
In color picker, if a color is selected, it is displayed at the "top", as soon as I tick/untick reverse intensity it becomes white. This is not the case when I select a lut.

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/agents/metadata/rnd/RendererControl.java:
```diff
@@ -506,17 +492,17 @@ public void propertyChange(PropertyChangeEvent evt)
      */
     private void handleColorPicker(boolean reset, boolean preview, int index,
-            Color newColor, Color oldColor, String newLut, String oldLut) {
+            Color newColor, Color oldColor, String newLut, String oldLut, boolean revInt) {
```
new parameter to add to javadoc

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/env/rnd/PixelsServicesFactory.java:
```diff
@@ -145,6 +147,15 @@ public static RndProxyDef convert(RenderingDef rndDef)
+                
+                cb.setReverseIntensity(false);
+                List<CodomainMapContext> cdctx = c.copySpatialDomainEnhancement();
+                for(CodomainMapContext cd : cdctx) {
```
usually we have space between for/if and bracket

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/env/rnd/RenderingControl.java:
```diff
@@ -877,4 +877,25 @@ void setLookupTable(int w, String lut) throws RenderingServiceException,
+     *            The channel index
+     * @return See above
+     */
+    boolean getReverseIntensity(int w);
```
maybe use `index` as in the other method

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/env/rnd/RenderingControl.java:
```diff
@@ -877,4 +877,25 @@ void setLookupTable(int w, String lut) throws RenderingServiceException,
+     * @param revInt
+     *            The reverse intensity flag
+     */
+    void setReverseIntensity(int w, boolean revInt)
```
maybe use `index` as in the other method

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/env/rnd/RenderingControlProxy.java:
```diff
@@ -2120,6 +2151,61 @@ public String getLookupTable(int w) {
+     * @see RenderingControl#getReverseIntensity(int)
+     */
+    @Override
+    public boolean getReverseIntensity(int w) {
```
same `index`

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/env/rnd/RenderingControlProxy.java:
```diff
@@ -2120,6 +2151,61 @@ public String getLookupTable(int w) {
+     * @see RenderingControl#setReverseIntensity(int, boolean)
+     */
+    @Override
+    public void setReverseIntensity(int w, boolean revInt)
```
same

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/util/ui/colourpicker/ColourObject.java:
```diff
@@ -22,77 +22,110 @@
 package org.openmicroscopy.shoola.util.ui.colourpicker;
 
-
 //Java imports
```
remove //java imports and the other 2

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/util/ui/colourpicker/ColourPicker.java:
```diff
@@ -139,46 +131,73 @@ void cancel()
+        newColor.lut = model.getLUT();
+        oldColor.lut = model.getOriginalLUT();
+        
+        newColor.revInt = model.getReversetIntensity();
```
method name has an extra `t`. The method name needs to be fixed.

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/util/ui/colourpicker/ColourPicker.java:
```diff
@@ -251,9 +271,9 @@ public ColourPicker(JFrame owner, Color color, boolean field) {
      * @param selectedLUT The selected lookup table
      */
-    public ColourPicker(JFrame owner, Color color, Collection<String> luts, String selectedLUT)
+    public ColourPicker(JFrame owner, Color color, Collection<String> luts, String selectedLUT, boolean revInt)
```
javadoc missing new param

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/util/ui/colourpicker/TabbedPaneUI.java:
```diff
@@ -407,9 +422,22 @@ String getDescription()
 		String text = fieldDescription.getText();
 		if (text == null) return null;
-		return text.trim();
+		text = text.trim();
```
`StringUtils.isBlank` will handle `null` and `length=0`

dominikl commented:
Thanks @jburel . All issues should be fixed now. Will add it to re-review tomorrow.

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/env/rnd/PixelsServicesFactory.java:
```diff
@@ -150,7 +150,7 @@ public static RndProxyDef convert(RenderingDef rndDef)
                 List<CodomainMapContext> cdctx = c.copySpatialDomainEnhancement();
-                for(CodomainMapContext cd : cdctx) {
+                for (CodomainMapContext cd : cdctx) {
                     if(cd instanceof ReverseIntensityContext) {
```
space before `if`

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/env/rnd/RenderingControl.java:
```diff
@@ -411,45 +411,36 @@ public void setActive(int w, boolean active)
-     * @param w The channel
+     * @param index The channel
      * @return See above.
+     * @throws RenderingServiceException    If an error occurred while getting 
```
while retrieving the contexts

dominikl replied:
Would it make sense to just say 'If an error occurred', the 'while ...' part basically just repeats what is said in the method description anyway?

jburel replied:
makes sense

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/env/rnd/RenderingControl.java:
```diff
@@ -839,6 +831,9 @@ public Dimension getTileSize()
 	 * Returns the list of the levels.
 	 * 
 	 * @return See above.
+	 * @throws RenderingServiceException   If an error occurred while getting 
```
while retrieving the values

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/util/ui/colourpicker/TabbedPaneUI.java:
```diff
@@ -52,6 +52,8 @@
 //Third-party libraries
 
+
+import org.apache.commons.lang.StringUtils;
```
same here could remove //Third-party etc.

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/util/ui/colourpicker/TabbedPaneUI.java:
```diff
@@ -421,9 +421,7 @@ String getDescription()
-		if(text.length()==0)
+		if (StringUtils.isBlank(text))
 		    return null;
 		return text;
```
I will still return `text.trim()`

dominikl commented:
Last commit is javadoc only, no functional implication, so testing can still go ahead today if needed.

jburel commented:
Problem noticed has been fixed and changes in doc done.
Merging Thanks
